### ISSUE TITLE
Question regarding shuffling data before scaffold splitting?

### ISSUE THREAD

**Issue by NanyBeaSun** [thaonguyen217/farm_molecular_representation#1]
Hello, I have read your code in downstream_tasks/(0)scaffold_split.py, and I would like to ask why the data is shuffled before scaffold splitting? I have read other codes that also use deepchem for scaffold splitting, and it seems that they do not shuffle the data, thereby ensuring that the data remains non-random.

like:
https://github.com/chao1224/GraphMVP/blob/main/src_classification/splitters.py
https://github.com/idrugLab/FG-BERT/blob/main/dataset_scaffold_random.py
```python
    if args.split == 'scaffold':
        smiles_list = pd.read_csv(dataset_folder + args.dataset + '/processed/smiles.csv',
                                  header=None)[0].tolist()
        train_dataset, valid_dataset, test_dataset = scaffold_split(
            dataset, smiles_list, null_value=0, frac_train=0.8,
            frac_valid=0.1, frac_test=0.1)
        print('split via scaffold')
```

```python


def scaffold_split(dataset, smiles_list, task_idx=None, null_value=0,
                   frac_train=0.8, frac_valid=0.1, frac_test=0.1,
                   return_smiles=False):
    """
    Adapted from https://github.com/deepchem/deepchem/blob/master/deepchem/splits/splitters.py
    Split dataset by Bemis-Murcko scaffolds
    This function can also ignore examples containing null values for a
    selected task when splitting. Deterministic split
    :param dataset: pytorch geometric dataset obj
    :param smiles_list: list of smiles corresponding to the dataset obj
    :param task_idx: column idx of the data.y tensor. Will filter out
        examples with null value in specified task column of the data.y tensor
        prior to splitting. If None, then no filtering
    :param null_value: float that specifies null value in data.y to filter if
        task_idx is provided
    :param frac_train, frac_valid, frac_test: fractions
    :param return_smiles: return SMILES if Ture
    :return: train, valid, test slices of the input dataset obj. """
    np.testing.assert_almost_equal(frac_train + frac_valid + frac_test, 1.0)

    if task_idx is not None:
        # filter based on null values in task_idx
        # get task array
        y_task = np.array([data.y[task_idx].item() for data in dataset])
        # boolean array that correspond to non null values
        non_null = y_task != null_value
        smiles_list = list(compress(enumerate(smiles_list), non_null))
    else:
        non_null = np.ones(len(dataset)) == 1
        smiles_list = list(compress(enumerate(smiles_list), non_null))

    # create dict of the form {scaffold_i: [idx1, idx....]}
    all_scaffolds = {}
    for i, smiles in smiles_list:
        scaffold = generate_scaffold(smiles, include_chirality=True)
        if scaffold not in all_scaffolds:
            all_scaffolds[scaffold] = [i]
        else:
            all_scaffolds[scaffold].append(i)

    # sort from largest to smallest sets
    all_scaffolds = {key: sorted(value) for key, value in all_scaffolds.items()}
    all_scaffold_sets = [
        scaffold_set for (scaffold, scaffold_set) in sorted(
            all_scaffolds.items(), key=lambda x: (len(x[1]), x[1][0]), reverse=True)
    ]

    # get train, valid test indices
    train_cutoff = frac_train * len(smiles_list)
    valid_cutoff = (frac_train + frac_valid) * len(smiles_list)
    train_idx, valid_idx, test_idx = [], [], []
    for scaffold_set in all_scaffold_sets:
        if len(train_idx) + len(scaffold_set) > train_cutoff:
            if len(train_idx) + len(valid_idx) + len(scaffold_set) > valid_cutoff:
                test_idx.extend(scaffold_set)
            else:
                valid_idx.extend(scaffold_set)
        else:
            train_idx.extend(scaffold_set)

    assert len(set(train_idx).intersection(set(valid_idx))) == 0
    assert len(set(test_idx).intersection(set(valid_idx))) == 0

    train_dataset = dataset[torch.tensor(train_idx)]
    valid_dataset = dataset[torch.tensor(valid_idx)]
    test_dataset = dataset[torch.tensor(test_idx)]

    if not return_smiles:
        return train_dataset, valid_dataset, test_dataset
    else:
        train_smiles = [smiles_list[i][1] for i in train_idx]
        valid_smiles = [smiles_list[i][1] for i in valid_idx]
        test_smiles = [smiles_list[i][1] for i in test_idx]
        return train_dataset, valid_dataset, test_dataset, \
               (train_smiles, valid_smiles, test_smiles)
```

I saw another type of data splitting method where they typically use chemprop for scaffold splitting, such as in KANO:
https://github.com/HICAI-ZJU/KANO/blob/main/chemprop/data/scaffold.py
When the "balanced" parameter is set to true, it introduces randomness.
```python
...
    elif split_type == 'scaffold_balanced':
        return scaffold_split(data, sizes=sizes, balanced=True, seed=seed, 
                              logger=logger)
...
```
```python

def scaffold_split(data: MoleculeDataset,
                   sizes: Tuple[float, float, float] = (0.8, 0.1, 0.1),
                   balanced: bool = False,
                   seed: int = 0,
                   logger: logging.Logger = None) -> Tuple[MoleculeDataset,
                                                           MoleculeDataset,
                                                           MoleculeDataset]:
    """
    Split a dataset by scaffold so that no molecules sharing a scaffold are in the same split.

    :param data: A MoleculeDataset.
    :param sizes: A length-3 tuple with the proportions of data in the
    train, validation, and test sets.
    :param balanced: Try to balance sizes of scaffolds in each set, rather than just putting smallest in test set.
    :param seed: Seed for shuffling when doing balanced splitting.
    :param logger: A logger.
    :return: A tuple containing the train, validation, and test splits of the data.
    """
    assert sum(sizes) == 1

    # Split
    train_size, val_size, test_size = sizes[0] * len(data), sizes[1] * len(data), sizes[2] * len(data)
    train, val, test = [], [], []
    train_scaffold_count, val_scaffold_count, test_scaffold_count = 0, 0, 0

    # Map from scaffold to index in the data
    scaffold_to_indices = scaffold_to_smiles(data.mols(), use_indices=True)

    if balanced:  # Put stuff that's bigger than half the val/test size into train, rest just order randomly
        index_sets = list(scaffold_to_indices.values())
        big_index_sets = []
        small_index_sets = []
        for index_set in index_sets:
            if len(index_set) > val_size / 2 or len(index_set) > test_size / 2:
                big_index_sets.append(index_set)
            else:
                small_index_sets.append(index_set)
        random.seed(seed)
        random.shuffle(big_index_sets)
        random.shuffle(small_index_sets)
        index_sets = big_index_sets + small_index_sets
    else:  # Sort from largest to smallest scaffold sets
        index_sets = sorted(list(scaffold_to_indices.values()),
                            key=lambda index_set: len(index_set),
                            reverse=True)

    for index_set in index_sets:
        if len(train) + len(index_set) <= train_size:
            train += index_set
            train_scaffold_count += 1
        elif len(val) + len(index_set) <= val_size:
            val += index_set
            val_scaffold_count += 1
        else:
            test += index_set
            test_scaffold_count += 1

    if logger is not None:
        logger.debug(f'Total scaffolds = {len(scaffold_to_indices):,} | '
                     f'train scaffolds = {train_scaffold_count:,} | '
                     f'val scaffolds = {val_scaffold_count:,} | '
                     f'test scaffolds = {test_scaffold_count:,}')
    
    log_scaffold_stats(data, index_sets, logger=logger)

    # Map from indices to data
    train = [data[i] for i in train]
    val = [data[i] for i in val]
    test = [data[i] for i in test]

    return MoleculeDataset(train), MoleculeDataset(val), MoleculeDataset(test)
```

I reviewed some literature and found this source: https://arxiv.org/html/2311.02798v3, which discusses the two types of scaffold splitting methods mentioned above, categorized as deterministic scaffold splits and balanced scaffold splits.

> Note that the reported results of GROVER, KPGT, MoLFormer and KANO in their original papers are evaluated under the balanced scaffold splits, which is different than the deterministic scaffold splits used in this work and the rest of the baseline methods. Both splits aim to hold out a set of molecules with scaffolds that are not seen during training. The deterministic scaffold splits prioritizes the satisfaction of the specified split ratio, while the balanced scaffold splits prioritizes the balance of scaffold frequency across the training, validation, and test sets. We use the deterministic scaffold splits in our experiments primarily because most of the baselines also employ it. All other aspects, including hyperparameter choices for both the models and the training setup, remain consistent with the default configurations of these baseline methods for replicating the results.

In https://github.com/yuewan2/MolMCL/issues/1, the author mentioned:

> Datasets from MoleculeNet (including BBBP) are splitted by molmcl.splitters.scaffold_split(). It is a deterministic split that does not require random seed. Therefore, the resulted split from this function should be exactly the same as what's used in the paper. 

Therefore, I believe that if deepchem's approach is used for data splitting, it should ensure that the resulting dataset is consistent every time.

Your code uses deepchem for data splitting, which should fall under deterministic scaffold splits. However, shuffling the data before calling the scaffold splitting function changes the relative positions of scaffolds with the same frequency, resulting in different train-val-test splits each time.

My understanding of scaffold splitting techniques is very limited. Based on the above, I think that before performing scaffold splitting in your code, the dataset should not be shuffled. I wonder if my understanding is incorrect.  
If you have time, I hope you can help clarify my confusion. Thank you very much.

**Comment by mxqmxqmxq** [comment]
Hello, regarding this question, I can share some suggestions:
In most cases, we use scaffold split with balanced set to False. This corresponds to a stricter and more deterministic splitting strategy, ensuring consistent results across runs.
If balanced is set to True, randomness is introduced, the splits are no longer deterministic, and the test set generally becomes easier.
